libgalois/src/PropertyGraph.cpp

```diff
@@ -145,6 +146,182 @@ MakePropertyGraph(
       std::move(rdg_file), std::move(rdg_result.value()));
 }
 
+/// Assumes all boolean or uint8 properties are types
+katana::Result<std::shared_ptr<arrow::NumericArray<arrow::UInt8Type>>>
+GetTypeIDsFromProperties(
+    const std::shared_ptr<arrow::Table>& properties,
+    katana::PropertyGraph::TypeToTypeIDsMap* type_name_to_ids,
+    katana::PropertyGraph::TypeIDToTypesMap* type_id_to_names) {
+  // throw an error if each column/property has more than 1 chunk
+  for (int i = 0, n = properties->num_columns(); i < n; i++) {
+    std::shared_ptr<arrow::ChunkedArray> property = properties->column(i);
+    if (property->num_chunks() != 1) {
+      return KATANA_ERROR(
+          katana::ErrorCode::NotImplemented,
+          "execution failed because property {} has {} chunks (1 chunk "
+          "expected)",
+          properties->schema()->field(i)->name(), property->num_chunks());
+    }
+  }
+
+  // collect the list of types
+  std::vector<int> type_field_indices;
+  const std::shared_ptr<arrow::Schema>& schema = properties->schema();
+  KATANA_LOG_DEBUG_ASSERT(schema->num_fields() == properties->num_columns());
+  for (int i = 0, n = schema->num_fields(); i < n; i++) {
+    const std::shared_ptr<arrow::Field>& current_field = schema->field(i);
+
+    // a bool or uint8 property is (always) considered a type
+    // TODO(roshan) make this customizable by the user
+    if (current_field->type()->Equals(arrow::boolean()) ||
+        current_field->type()->Equals(arrow::uint8())) {
+      type_field_indices.push_back(i);
+    }
+  }
+
+  // assign a new ID to each type
+  // NB: cannot use unordered_map without defining a hash function for vectors;
+  // performance is not affected here because the map is very small (<=256)
+  std::map<std::vector<int>, katana::PropertyGraph::TypeID>
```

Comment:
(1) Lexicographic ordering for comparison is sufficient in this case. There's an external order we're enforcing for all the vector-keys (based on the order of columns in the properties table).

(2) Dense bit-set was the old code. It would need more memory, not less. Bit-set is suitable when any combination of distinct types can occur but this is almost never the case in database schemas. With 8 bits in a bit-set, we can only support 8 distinct types and 2^8 distinct combination of these distinct types. On the other hand, an arbitrary mapping like TypeID here can support 2^8 distinct "types or combinations of distinct types" in 8 bits. This is more than sufficient for most datasets. If needed, we can use 16 bits instead of 8 bits - I don't see a need for anything more than that even for the most complex database schemas. With dynamic or sparse bit-set, 8 or 16 bits are insufficient, so we cannot use less than the memory of TypeID for all our datasets.

Reply:
I agree with what you said above. My comment is more so applicable to places where you have vectors (or sets) of typeIDs and sets(and vectors) of field indices. We keep the typeIDs as you've designed them and 8 bits is enough. When it comes to represent a set of typeIDs (or field indices), we can use a bitset (sparse or dense) because these IDs are from a linear space of non-negative integers, which makes perfect sense for bit-sets. 

A related question that I have is (for my understanding primarily): what is our flow of execution when we visit a node that has multiple types. If I understand, under this design, we'll assign a 8-bit ID to a unique combination of types applicable to this node. How do we look up what those types are?

Reply:
Now I understand your comment. Yes, `std::vector<bool>` may be better than `std::set<uint8_t>`.  Note that we don't need atomic accesses so `DynamicBitset` is not required. Also note that the size for none of these vectors or sets are O(vertices) or O(edges), so the memory overhead is not high in either case. Nevertheless, I think using `std::vector<bool>` will make the code simpler. I'll make a pass and try to update it everywhere. If there are any issues, I'll let you know.

I'll answer your flow question in the enterprise-repo because there's more context there.

Reply:
I decided to go with `std::bitset` for set of TypeIDs. However, for set of field indices (used only in this function), I decided to keep it as a vector of field indices (`katana::gstl::Vector<int>`) for these reasons:
(1) `std::bitset` does not support dynamic (fixed) size. The number of field indices is only known at runtime and I don't want to put any restriction on it because it includes properties.
(2) `std::vector<bool>` does not  support counting the set bits, so it would make this function more complex to do that.
(3) `katana::DynamicBitset` does not support thread-specific allocators, so it may make the function slower.

Reply:
`std::bitset` is a good choice for now. We should think about whether there's benefit to using sparse bit set as we expect sparsity when the graph has different kinds of nodes. 

Secondly, I don't see why using `std::bitset` with max number of bits is any worse than using `std::vector`. Similarly, I don't see (and you can help me understand) why `std::vector<bool>` doesn't work equally well.

Reply:
That's (2): `std::vector<bool>` does not support counting the set bits directly. It makes the code more verbose and difficult to read IMO.